File: code/data_structures/stack/stack/stack.js
/* Part of Cosmos by OpenGenus Foundation */

/* Stack!!
* A stack is exactly what it sounds like. An element gets added to the top of
* the stack and only the element on the top may be removed. This is an example
* of an array implementation of a Stack. So an element can only be added/removed
* from the end of the array.
*/

// Functions: push, pop, peek, view, length

//Creates a stack
var Stack = function () {
  //The top of the Stack
  this.top=0;
  //The array representation of the stack
  this.stack = {};

  //Adds a value onto the end of the stack
  this.push=function(value) {
    this.stack[this.top]=value;
    this.top++;
  }

  //Removes and returns the value at the end of the stack
  this.pop = function(){
    if(this.top === 0){
      return "Stack is Empty";
    }

    this.top--;
    var result = this.stack[this.top];
    delete this.stack[this.top];
    return result;
  }

  //Returns the size of the stack
  this.size = function(){
    return this.top;
  }

  //Returns the value at the end of the stack
  this.peek = function(){
    return this.stack[this.top-1];
  }

  //To see all the elements in the stack
  this.view= function(){
    for(var i=0;i<this.top;i++)
      console.log(this.stack[i]);
  }
}

//Implementation
var myStack = new Stack();

myStack.push(1);
myStack.push(5);
myStack.push(76);
myStack.push(69);
myStack.push(32);
myStack.push(54);
console.log(myStack.size());
console.log(myStack.peek());
console.log(myStack.pop());
console.log(myStack.peek());
console.log(myStack.pop());
console.log(myStack.peek());
myStack.push(55);
console.log(myStack.peek());
myStack.view();

File: algorithms/data_structures/stack/stack/stack.cs
/**
 * Stack implementation using a singly-linked link.
 * Part of the OpenGenus/cosmos project. (https://github.com/OpenGenus/cosmos)
 * 
 * A stack is a first-in first-out (FIFO) data structure.
 * Elements are manipulated by adding and removing elements of the top of the stack.
 */
using System;

namespace Cosmos_Data_Structures
{
    public class Stack<T>
    {
        //Node is a element that holds the data of the current element plus a reference to the next element.
        //Used to implement the singly-linked list.
        private class Node
        {
            public T data;
            public Node next;

            public Node(T data, Node next)
            {
                this.data = data;
                this.next = next;
            }
        }

        private Node top;
        public int Size { get; private set; }

        public Stack()
        {
            top = null;
            Size = 0;
        }

        //Add element to the top of the stack.
        public void Push(T element)
        {
            var newNode = new Node(element, top);
            top = newNode;
            Size++;
        }

        //Gets element at the top of the stack.
        //Throws an exception if the stack is empty.
        public T Peek()
        {
            if(IsEmpty())
            {
                throw new InvalidOperationException("Cannot peek on an empty stack!");
            }

            return top.data;
        }

        //Removes and returns the element at the top of the stack.
        //Throws an exception if the stack is empty.
        public T Pop()
        {
            if(IsEmpty())
            {
                throw new InvalidOperationException("Cannot pop on an empty stack!");
            }

            var oldTop = top;
            top = top.next;
            Size--;
            return oldTop.data;
        }

        //Returns true if stack contains no elements, false otherwise.
        public bool IsEmpty()
        {
            return top == null;
        }

        //Returns a string representation of the stack.
        public override string ToString()
        {
            Node tmp = top;
            string result = "Stack([Top] ";

            while(tmp != null)
            {
                result += tmp.data;
                tmp = tmp.next;

                if(tmp != null)
                {
                    result += " -> ";
                }
            }

            result += ")";
            return result;
        }
    }

    //Stack testing methods/class.
    public class StackTest
    {
        static void Main(string[] args)
        {
            Console.Write("Creating stack...");
            var intStack = new Stack<int>();

            for (int i = 0; i < 10; i++)
            {
                intStack.Push(i);
            }
            Console.WriteLine("done");

            Console.WriteLine(intStack.ToString());
            Console.WriteLine("Size of stack: " + intStack.Size);
            Console.WriteLine("Topmost element is " + intStack.Peek() + ".\n");

            Console.Write("Removing elements...");
            for (int i = 0; i < 3; i++)
            {
                intStack.Pop();
            }
            Console.WriteLine("done");

            Console.WriteLine(intStack.ToString());
            Console.WriteLine("Size of stack: " + intStack.Size);
            Console.WriteLine("Topmost element is " + intStack.Peek() + ".\n");

            Console.WriteLine("Press any key to continue.");
            Console.ReadKey();
        }
    }
}
